core/arch/atomic_exports.h

```diff
@@ -723,7 +723,296 @@ atomic_dec_becomes_zero(volatile int *var)
                                  : "r2")
 #        define SET_IF_NOT_ZERO(flag) SET_FLAG(ne, flag)
 #        define SET_IF_NOT_LESS(flag) SET_FLAG(ge, flag)
-#    endif /* X86/ARM */
+#    elif defined(DR_HOST_RISCV64)
+#        define ATOMIC_1BYTE_READ(addr_src, addr_res)               \
+            do {                                                    \
+                /* Fence for acquire semantics. */                  \
+                __asm__ __volatile__("lbu %0, 0(%1)\n\t"            \
+                                     "fence iorw,iorw"              \
+                                     :                              \
+                                     : "r"(addr_src), "r"(addr_res) \
+                                     : "memory");                   \
+            } while (0)
+#        define ATOMIC_1BYTE_WRITE(target, value, hot_patch)   \
+            do {                                               \
+                /* Not currently used to write code */         \
+                ASSERT_CURIOSITY(!hot_patch);                  \
+                /* Fence for release semantics. */             \
+                __asm__ __volatile__("fence iorw,iorw\n\t"     \
+                                     "sb %0, 0(%1)\n\t"        \
+                                     :                         \
+                                     : "r"(value), "r"(target) \
+                                     : "memory");              \
+            } while (0)
+#        define ATOMIC_4BYTE_ALIGNED_WRITE(target, value, hot_patch) \
+            do {                                                     \
+                /* Page 25 of Volume I: RISC-V Unprivileged ISA      \
+                 * V20191213 states that aligned loads/stores are    \
+                 * atomic. Alternatively we could use                \
+                 * amoswap.w zero %0, 0(%1) but that would incur     \
+                 * an extra read we don't need and also requires     \
+                 * alignment.                                        \
```

Comment:
We do have alignment for this one: as the name says `ATOMIC_4BYTE_ALIGNED_WRITE`.

With `riscv64-linux-gnu-g++` version 11.2.0, for:
```
std::atomic<int> myatomic;
myatomic.store(1, std::memory_order_release);
```

It does use `amoswap`:
```
 822:   0f50000f                fence   iorw,ow
 826:   08e7a02f                amoswap.w       zero,a4,(a5)
```

Is `amoswap` in every RSIC-V or only if atomics are specified?  We're targeting +A though so maybe it doesn't matter here.

Reply:
`amoswap` is part of "atomic instructions" extension but I am targeting A we could use it.
What I meant by this comment is that 4 byte store is only atomic in case of aligned stores (as mentioned in the spec), therefore using `amoswap` does not give us any benefit because it also needs an aligned address (which we have) and potentially loads a value from the updated memory location. Given that the destination register is `zero` (so hard-wired zero) I guess whether the load actually takes place or not depends on particular HW implementation.
Also I've found a record in GCC's bugzilla which notes that for `atomic_store_explicit` with `relaxed` memory order can be implemented by just a store: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=89835. That said I could use the same implementation as GCC's stdatomics if you prefer.

Also an interesting side-note is that the fence in the assembly you've posted above ensures that no writes to normal or I/O memory cannot be reordered before the fence (`fence iorw,ow`). My implementation also keeps the reads from reordering (`fence iorw,iorw`).

Reply:
Let's go with what you have.